### PR TITLE
add WITHCOUNT flag to TOPK.LIST 

### LINKED ISSUE #324: Need a command to return all TOP-K entries with their counts
Now I could do this using 2 queries(TOPK.LIST + TOPK.COUNT) but while doing so some items could have been expelled which is non ideal...

### PR DESCRIPTION
**Pull request by ashtul**
fixes #324

### CODE REVIEW

**Review by oshadmi — COMMENTED** [review]
Few questions

File: src/rm_topk.c [hunk 1]
```diff
@@ -187,25 +187,40 @@ static int TopK_Count_Cmd(RedisModuleCtx *ctx, RedisModuleString **argv, int arg
...
     }
-    uint32_t k = topk->k;
-    char **heapList = TOPK_CALLOC(k, (sizeof(char *)));
-    TopK_List(topk, heapList);
-    RedisModule_ReplyWithArray(ctx, k);
-    for (int i = 0; i < k; ++i) {
-        if (heapList[i] != NULL) {
-            RedisModule_ReplyWithSimpleString(ctx, heapList[i]);
+    HeapBucket *heapList = TopK_List(topk);
+    RedisModule_ReplyWithArray(ctx, REDISMODULE_POSTPONED_ARRAY_LEN);
```
Comment (filipecosta90) [inline src/rm_topk.c:209]: @ashtul given we know there are implications on `REDISMODULE_POSTPONED_ARRAY_LEN` and that we know beforehand the size of the array ( either k or 2xk ) lets keep RedisModule_ReplyWithArray with the proper arraylen. 
agree?

File: docs/TopK_Commands.md [hunk 2]
```diff
@@ -184,12 +184,13 @@ TOPK.COUNT test foo 42 nonexist
...
-TOPK.LIST {key}
+TOPK.LIST {key} [WITHCOUNT]
 ```
 
 ### Parameters
 
 * **key**: Name of sketch where item is counted.
+* **WITHCOUNT**: Count of each element is returned.  
 
 ### Complexity
```
Comment (filipecosta90) [inline docs/TopK_Commands.md:196]: @ashtul we're missing the changes to the reply. 
Also, we need to propagate the awareness of this change into all clients.

**Comment by filipecosta90** [comment]
@ashtul this a good opportunity to add a topk benchmark variation. 
Added some other comments about changes that IMHO can be avoided, and some missing docs.

File: src/rm_topk.c (hunk 1 above)
Comment (ashtul) [reply src/rm_topk.c:209]: We might break with less than `k` elements if the heap is not full yet.

File: src/rm_topk.c [hunk 3]
```diff
@@ -187,25 +187,40 @@ static int TopK_Count_Cmd(RedisModuleCtx *ctx, RedisModuleString **argv, int arg
...
+    long arrlen = 0;
+    for (int i = 0; i < topk->k; ++i) {
+        if (heapList[i].count != 0) {
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/rm_topk.c:212]: Before we had `char **` so `heapList[i]` held a pointer. Now we hold a struct that cannot be `NULL`.
This is why `cmpHeapBucket` is safe.

File: docs/TopK_Commands.md (hunk 2 above)
Comment (ashtul) [reply docs/TopK_Commands.md:196]: We don't always put output for all optional params as it makes the docs very long.

**Review by ashtul — COMMENTED** [review]
Comments answered or fixed.

File: src/rm_topk.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (oshadmi) [reply src/rm_topk.c:212]: If command is as before, without `WITHCOUNT`, would sorting and changing the order be a breaking change?
We could keep the previous behavior in this case (use pointers, do not sort, and avoid the REDISMODULE_POSTPONED_ARRAY_LEN) and also have better performance,
and only have the new behavior with `WITHCOUNT`.

[the remark this answers is not in the document]
Comment (ashtul) [reply src/rm_topk.c:212]: @oshadmi Changing the order should not be considered breaking change IMO.

File: src/rm_topk.c (hunk 1 above)
Comment (ashtul) [reply src/rm_topk.c:209]: @filipecosta90 a single use of `REDISMODULE_POSTPONED_ARRAY_LEN` would not hurt performance. From what I know, we had an issue in search where we had wide use of it in a reply true of possibly millions of results.
The other way to do it is to iterate twice over the array and count the legal result but this seems excessive to me. WDYT?

**Comment by filipecosta90** [comment]
@ashtul can we take this opportunity to add a benchmark definition for TOPK.LIST? 
Like we have for bf.add https://github.com/RedisBloom/RedisBloom/blob/master/tests/benchmarks/bf_add_cap10M_err0.001.yml

**Review by oshadmi — APPROVED** [review]
LGTM 👍🏼

**Review by filipecosta90 — APPROVED** [review]
LGTM. Thank you for the extra effort put into adding more benchmarks to this!

**Pull request merged**